core/src/services/s3/error.rs

```diff
@@ -41,7 +41,7 @@ pub fn parse_error(resp: Response<Buffer>) -> Error {
     let (mut kind, mut retryable) = match parts.status.as_u16() {
         403 => (ErrorKind::PermissionDenied, false),
         404 => (ErrorKind::NotFound, false),
-        304 | 412 => (ErrorKind::ConditionNotMatch, false),
+        304 | 400 | 412 => (ErrorKind::ConditionNotMatch, false),
```

Comment:
This could lead to incorrect behavior as we encounter other 400 errors. Which specific error do you want to address?

Reply:
when performing a `stat`/`read`/`delete` request with a not existing version, we will get `400 InvalidArgument` error.
```
Context:
   uri: http://127.0.0.1:9000/first-bucket/6876d14b-de9d-41c4-81fe-77f47315890a?versionId=not-existing-version
   response: Parts { status: 400, version: HTTP/1.1, headers: {"accept-ranges": "bytes", "content-length": "0", "server": "MinIO", "strict-transport-security": "max-age=31536000; includeSubDomains", "vary": "Origin", "vary": "Accept-Encoding", "x-amz-bucket-region": "local", "x-amz-id-2": "dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8", "x-amz-request-id": "17F8398D1704F108", "x-content-type-options": "nosniff", "x-minio-error-code": "InvalidArgument", "x-minio-error-desc": "\"Invalid version id specified\"", "x-xss-protection": "1; mode=block", "date": "Tue, 24 Sep 2024 16:02:15 GMT"} }
   service: s3
   path: 6876d14b-de9d-41c4-81fe-77f47315890a
```

Reply:
This error shouldn't be treated as `ConditionNotMatch` because there's no condition to match. I prefer to return `ErrorKind::NotFound` in this case, but I think we need to check more than just the status code, as it's possible for us to confuse it with other 400 errors.

Reply:
In S3, it's `400` too, and there is no error message.
```
Context:
   uri: https://s3.us-east-2.amazonaws.com/logan-first-bucket/5650e908-5fce-4e2b-b692-47d57ddc2696/853931a9-6f0a-453a-a156-aaf54a08b1e5?versionId=not-existing-version
   response: Parts { status: 400, version: HTTP/1.1, headers: {"x-amz-request-id": "BAVB32C46K693YXH", "x-amz-id-2": "1gVzFxd4m5QC/amSUDg5kq3F3XzdKAIjK4RaSktM27SpQVeh38VDsEgaJK6cDTzGbSEhLFtXrUI=", "content-type": "application/xml", "transfer-encoding": "chunked", "date": "Tue, 24 Sep 2024 16:18:54 GMT", "connection": "close", "server": "AmazonS3"} }
   service: s3
   path: 853931a9-6f0a-453a-a156-aaf54a08b1e5
```

Reply:
Hi, have you tried a valid but not exist version id?

Reply:
> This error shouldn't be treated as `ConditionNotMatch` because there's no condition to match. 

I don't get this.  In a versioned bucket, performing a `read` request with a not-existent version still results in  `400`. could we treat it as a `ConditionNotMatch` error ?

Reply:
> Hi, have you tried a valid but not exist version id?

Oh, it'll return `404`.

Reply:
> Oh, it'll return `404`.

Thanks a lot! That's what I expected to see.

Reply:
Unfortunately, `S3` and `Minio` use different version formats, meaning that a valid version in `S3` might not be valid in `Minio` :(

Reply:
> Unfortunately, `S3` and `Minio` use different version formats, meaning that a valid version in `S3` might not be valid in `Minio` :(

Let test in this way:

- Write `PathA`, get version `A`
- Write a new file with `PathB`
- Read/Stat/Delete `PathB` with version `A`

Reply:
It seems that `Delete` behaves differently. Deleting a not-existent version doesn't return an error in `Minio` or `S3`, and Deleting a not-existent file doesn't return an error In `OpenDAL` now. So, how about ensuring that deleting an not-existent version always returns `Ok` ?

Reply:
> It seems that `Delete` behaves differently. Deleting a not-existent version doesn't return an error in `Minio` or `S3`, and Deleting a not-existent file doesn't return an error in `OpenDAL` now. So, how about ensuring that deleting an not-existent version always returns `Ok` ?

Yes, it's expected that removing a non-existent version returns `Ok`.